docs/Components/daemon.md

```diff
@@ -0,0 +1,149 @@
+system bus, and takes ownership of the name `org.mozilla.vpn.dbus`.
+
+The Linux kernel includes support for the Wireguard protocol, so this daemon only needs to respond to the communication
+protocol and retain the necessary capabilities to orchestrate interface creation, configuration and management.
+
+### MacOS launchd
+
+The MacOS daemon is launched using [launchd](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html)
+by installing a daemon plist at `/Library/LaunchDaemons/org.mozilla.macos.FirefoxVPN.daemon.plist` and registering the
+daemon for automatic startup with `launchctl load -w $DAEMON_PLIST_PATH`.
+
+Communication with the daemon is achieved via a named UNIX socket, which the daemon will create at
+`/var/run/mozillavpn/daemon.socket`. The API methods are invoked by formatting the arguments as a JSON object, and
+inserting a value of `type` set to the name of the method being invoked.
+
+MacOS does not include native support for the Wireguard protocol, so in order to establish a Wireguard tunnel, the
+daemon will create and manage other processes to perform this function. The
+[wireguard-go](https://git.zx2c4.com/wireguard-go) project provides a packet tunnel for Wireguard connectivity and
+connects to the kernel using the `/dev/net/tun` interface.
+
+To manage the DNS resolver changes required for this interface, we also include another tool, the `macosdnsmanager`,
+whose purpose is to reconfigure MacOS to use the DNS resolver provided by the VPN network, and to restore the previous
+DNS  configuration upon shutdown of the VPN tunnel.
+
+### Windows Service
+
+The MSI Installer package creates a Windows service, named the `Mozilla VPN (broker)`, to gain administrator privileges
+for creating and managing the network interfaces and configuration. This service is set to launch automatically on boot,
+and can be inspected via the Windows Service Manager.
+
+Communication with the daemon is achieved via a named pipe, which the daemon will create at `\\.\pipe\mozillavpn`.
+The API methods are invoked by formatting the arguments as a JSON object, and inserting a value of `type` set to the
+name of the method being invoked.
+
+Windows does not include native support for the Wireguard protocol, so in order provide this capability the
+[Wireguard NT](https://github.com/WireGuard/wireguard-nt) kernel driver is installed by the windows service. This driver
+is instantiated when bringing up the VPN tunnel to handle the connection.
+
+Additionally, a [split-tunneling](https://github.com/mullvad/win-split-tunnel) driver may also be provided, which can be used
+by the VPN to allow some applications to access the network directly and bypass the VPN. The daemon will check for this driver
+at startup, and register it for split tunneling if possible.
+
+### Android daemon???
+
+TODO: How does Android work, exactly?
```

Comment:
The Daemon is an [Android Service](https://developer.android.com/develop/background-work/services) communicating using [Bindings ](https://developer.android.com/develop/background-work/services/bound-services) with the Client (or widgets). 

The protocol is similar to the desktop daemons as we exchange json blobs but the format on the wire is `{requestType:int,data:json}` (...we could actually think about using the desktop protocol at some point)

To activate the daemon will request a raw handle of a TUN device using [Android's VPN API](https://developer.android.com/reference/android/net/VpnService.Builder). It load's wireguard-go into it's process and calls into wg-go with the handle and the config, which will then spawn a goroutine handling traffic.

Reply:
Thank you!